File: src/editor/gui/filter/editor/ColorFilterPanel.java
package editor.gui.filter.editor;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

import javax.swing.BoxLayout;
import javax.swing.JCheckBox;
import javax.swing.JLabel;

import editor.database.characteristics.ManaType;
import editor.database.symbol.ColorSymbol;
import editor.filter.Filter;
import editor.filter.FilterFactory;
import editor.filter.leaf.ColorFilter;
import editor.filter.leaf.FilterLeaf;
import editor.gui.generic.ComboBoxPanel;
import editor.util.Containment;

/**
 * This class represents a panel corresponding to a filter that groups
 * cards by a color characteristic.
 * 
 * @author Alec Roelke
 */
@SuppressWarnings("serial")
public class ColorFilterPanel extends FilterEditorPanel<ColorFilter>
{
	/**
	 * Map of colors onto their corresponding check boxes.
	 */
	private Map<ManaType, JCheckBox> colors;
	/**
	 * Combo box showing the containment options.
	 */
	private ComboBoxPanel<Containment> contain;
	/**
	 * Check box indicating that only multicolored cards should be matched.
	 */
	private JCheckBox multiCheckBox;
	/**
	 * Type of the filter being edited.
	 */
	private String type;
	
	/**
	 * Create a new ColorFilterPanel.
	 */
	public ColorFilterPanel()
	{
		super();
		setLayout(new BoxLayout(this, BoxLayout.X_AXIS));
		
		// Containment options
		contain = new ComboBoxPanel<>(Containment.values());
		add(contain);
		
		// Check boxes for selecting colors
		colors = new HashMap<>();
		for (ManaType color: ManaType.colors())
		{
			JCheckBox box = new JCheckBox();
			colors.put(color, box);
			add(box);
			JLabel symbol = new JLabel(ColorSymbol.SYMBOLS.get(color).getIcon(13));
			add(symbol);
		}
		
		// Check box for multicolored cards
		multiCheckBox = new JCheckBox("Multicolored");
		add(multiCheckBox);
	}
	
	/**
	 * Create a new ColorFilterPanel with initial contents obtained
	 * from the given filter.
	 * 
	 * @param f filter to get the contents from
	 */
	public ColorFilterPanel(ColorFilter f)
	{
		this();
		setContents(f);
	}

	@Override
	public Filter filter()
	{
		ColorFilter filter = (ColorFilter)FilterFactory.createFilter(type);
		filter.contain = contain.getSelectedItem();
		filter.colors.addAll(colors.keySet().stream().filter((c) -> colors.get(c).isSelected()).collect(Collectors.toSet()));
		filter.multicolored = multiCheckBox.isSelected();
		return filter;
	}

	@Override
	public void setContents(ColorFilter filter)
	{
		type = filter.type();
		contain.setSelectedItem(filter.contain);
		for (ManaType color: ManaType.colors())
			colors.get(color).setSelected(filter.colors.contains(color));
		multiCheckBox.setSelected(filter.multicolored);
	}

	@Override
	public void setContents(FilterLeaf<?> filter) throws IllegalArgumentException
	{
		if (filter instanceof ColorFilter)
			setContents((ColorFilter)filter);
		else
			throw new IllegalArgumentException("Illegal color filter " + filter.type());
	}
}

File: src/main/scala/editor/gui/filter/editor/ColorFilterPanel.scala
package editor.gui.filter.editor

import editor.database.attributes.CardAttribute
import editor.database.attributes.ManaType
import editor.database.symbol.ManaSymbolInstances.ColorSymbol
import editor.database.symbol.ManaSymbolInstances.StaticSymbol
import editor.filter.leaf.ColorFilter
import editor.filter.leaf.FilterLeaf
import editor.gui.filter.FilterSelectorPanel
import editor.gui.generic.ComboBoxPanel
import editor.gui.generic.ComponentUtils
import editor.util.Containment

import javax.swing.Box
import javax.swing.BoxLayout
import javax.swing.JCheckBox
import javax.swing.JLabel
import scala.collection.immutable.ListMap
import scala.jdk.CollectionConverters._

/**
 * Convenience constructors for creating [[ColorFilterPanel]]s.
 * @author Alec Roelke
 */
object ColorFilterPanel {
  /** @return an empty [[ColorFilterPanel]] set to filter by color */
  def apply(selector: FilterSelectorPanel) = new ColorFilterPanel(selector)

  /**
   * Create a new [[ColorFilterPanel]], set its filter attribute, and set its contents.
   * 
   * @param filter filter to use to set the new panel's attribute and contents
   * @return a new [[ColorFilterPanel]] with the given filter's attribute and contents
   */
  def apply(filter: ColorFilter, selector: FilterSelectorPanel) = {
    val panel = new ColorFilterPanel(selector)
    panel.setContents(filter)
    panel
  }
}

/**
 * A panel for customzing a filter that filters cards by color characteristic (e.g. color or color identity).
 * @author Alec Roelke
 */
class ColorFilterPanel(selector: FilterSelectorPanel) extends FilterEditorPanel[ColorFilter] {
  private val IconHeight = 13

  setLayout(BoxLayout(this, BoxLayout.X_AXIS))

  // Containment options
  private val contain = ComboBoxPanel(Containment.values)
  add(contain)

  // Check box for filtering for colorless
  private val colorless = JCheckBox()

  // Check boxes for selecting colors
  val colorBoxes = ListMap(ManaType.colors.map(_ -> JCheckBox()):_*)
  colorBoxes.foreach{ case (color, box) =>
    add(box)
    add(JLabel(ColorSymbol(color).getIcon(IconHeight)))
    box.addActionListener(_ => if (box.isSelected) colorless.setSelected(false))
  }
  add(Box.createHorizontalStrut(4))
  add(ComponentUtils.createHorizontalSeparator(4, contain.getPreferredSize.height))

  // Check box for multicolored cards
  private val multi = JCheckBox()
  add(multi)
  add(JLabel(StaticSymbol("M").getIcon(IconHeight)))
  multi.addActionListener(_ => if (multi.isSelected) colorless.setSelected(false))

  // Actually add the colorless box here
  colorless.setSelected(true)
  add(colorless)
  add(JLabel(ColorSymbol(ManaType.Colorless).getIcon(IconHeight)))
  colorless.addActionListener(_ => if (colorless.isSelected) {
    colorBoxes.foreach{ case (_, box) => box.setSelected(false) }
    multi.setSelected(false)
  })

  add(Box.createHorizontalStrut(2))

  protected override var attribute = CardAttribute.Colors

  override def filter = attribute.filter.copy(faces = selector.faces, contain = contain.getSelectedItem, colors = colorBoxes.collect{ case (c, b) if b.isSelected => c }.toSet, multicolored = multi.isSelected)

  override def setFields(filter: ColorFilter) = {
    attribute = filter.attribute
    contain.setSelectedItem(filter.contain)
    filter.colors.foreach(colorBoxes(_).setSelected(true))
    multi.setSelected(filter.multicolored)
    colorless.setSelected(!filter.multicolored && filter.colors.isEmpty)
  }
}